src/main/java/org/junitpioneer/jupiter/CartesianProductTestExtension.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2015-2020 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter;
+
+import static org.junit.platform.commons.support.AnnotationSupport.findAnnotation;
+import static org.junit.platform.commons.support.ReflectionSupport.invokeMethod;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContext;
+import org.junit.jupiter.api.extension.TestTemplateInvocationContextProvider;
+
+class CartesianProductTestExtension implements TestTemplateInvocationContextProvider {
+
+	@Override
+	public boolean supportsTestTemplate(ExtensionContext context) {
+		return findAnnotation(context.getTestMethod(), CartesianProductTest.class).isPresent();
```

Comment:
Bit of an overkill, don't you think?

Reply:
Then you argue to throw Pioneer Annotations out, because we can use the JUnit things. The idea was Pioneer Annotation was to encapsulate the JUnit thing with our own, improved methods. But I'll leave it to you / the others which way we use.

Same for TestKit or PioneerAssertions 🤷‍♂️

Reply:
The advantage of our annotation methods is that they make it easy to find meta annotations that are _indirectly present: (if a supertype of the element is annotated), _meta-present_ (if an annotation that is present on the element is itself annotated), or _enclosing-present (if an enclosing type [think opposite of `@Nested`] is annotated). As I see it, only _meta-present_ is relevant here, but that should indeed be allowed. Not sure if `AnnotationSupport.findAnnotation` does that.